### PR TITLE
r.out.gdal: add offset/scale options

### PR DESCRIPTION
**Pull request by landam**
This PR adds two new parameters to `r.out.gdal`:

```
     offset   Assign a specified offset value to output bands
      scale   Assign a specified scale value to output bands
```

Example:

```
r.out.gdal input=random output=random.tif offset=100 scale=0.01

gdalinfo random.tif
...
Band 1 Block=100x40 Type=UInt16, ColorInterp=Palette
  Description = random
    Computed Min/Max=1.000,3661.000
  NoData Value=65535
  Offset: 100,   Scale:0.01
...
```

### CODE REVIEW

File: raster/r.out.gdal/main.c [hunk 1]
```diff
@@ -586,6 +604,20 @@ int main(int argc, char *argv[])
 	    set_default_nodata_value(datatype, export_min, export_max);
     }
 
+    /* Offset value */
+    double *offsetval = NULL; /* not defined */
+    if (offsetopt->answer != NULL) {
+        offsetval = (double *) G_malloc(sizeof(double));
+        *offsetval = atof(offsetopt->answer);
```
Comment (metzm) [inline raster/r.out.gdal/main.c:611]:
```suggestion
    double offsetval = 0; /* not defined */
    if (offsetopt->answer != NULL) {
        offsetval = atof(offsetopt->answer);
```

File: raster/r.out.gdal/main.c [hunk 2]
```diff
@@ -586,6 +604,20 @@ int main(int argc, char *argv[])
...
+    }
+
+    /* Scale value */
+    double *scaleval = NULL; /* not defined */
+    if (scaleopt->answer != NULL) {
+        scaleval = (double *) G_malloc(sizeof(double));
+        *scaleval = atof(scaleopt->answer);
```
Comment (metzm) [inline raster/r.out.gdal/main.c:618]:
```suggestion
    double scaleval =1; /* not defined */
    if (scaleopt->answer != NULL) {
        scaleval = atof(scaleopt->answer);
```

File: raster/r.out.gdal/local_proto.h [hunk 3]
```diff
@@ -60,7 +60,7 @@
 /* export_band.c */
 int export_band(GDALDatasetH, int, const char *, 
 		const char *, struct Cell_head *, RASTER_MAP_TYPE, 
-		double, int, int, int);
+		double, int, int, int, double *, double *);
```
Comment (metzm) [inline raster/r.out.gdal/local_proto.h:63]:
```suggestion
		double, int, int, int, double, double);
```

File: raster/r.out.gdal/export_band.c [hunk 4]
```diff
@@ -215,7 +215,8 @@ int export_band(GDALDatasetH hMEMDS, int band,
 		const char *name, const char *mapset,
 		struct Cell_head *cellhead, RASTER_MAP_TYPE maptype,
 		double nodataval, int suppress_main_colortable, 
-		int no_metadata, int writenodata)
+		int no_metadata, int writenodata,
+                double *offsetval, double *scaleval)
```
Comment (metzm) [inline raster/r.out.gdal/export_band.c:219]:
```suggestion
                double offsetval, double scaleval)
```

File: raster/r.out.gdal/export_band.c [hunk 5]
```diff
@@ -355,6 +356,16 @@ int export_band(GDALDatasetH hMEMDS, int band,
 		GDALSetMetadataItem(hBand, key, value, NULL);
 	    }
 	}
+
+        /* apply offset factor if not defined */
+        if (offsetval) {
```
Comment (metzm) [inline raster/r.out.gdal/export_band.c:361]:
```suggestion
        /* apply offset factor if not 0 (zero) */
        if (offsetval) {
```

File: raster/r.out.gdal/export_band.c [hunk 6]
```diff
@@ -355,6 +356,16 @@ int export_band(GDALDatasetH hMEMDS, int band,
...
+            GDALSetRasterOffset(hBand, *offsetval);
```
Comment (metzm) [inline raster/r.out.gdal/export_band.c:362]:
```suggestion
        /* apply offset factor if not 0 (zero= */
        if (offsetval) {
            GDALSetRasterOffset(hBand, offsetval);
```

File: raster/r.out.gdal/export_band.c [hunk 7]
```diff
@@ -355,6 +356,16 @@ int export_band(GDALDatasetH hMEMDS, int band,
...
+        }
+
+        /* apply scale factor if not defined */
+        if (scaleval) {
+            GDALSetRasterScale(hBand, *scaleval);
```
Comment (metzm) [inline raster/r.out.gdal/export_band.c:367]:
```suggestion
        /* apply scale factor if not 1 */
        if (scaleval != 1) {
            GDALSetRasterScale(hBand, scaleval);
```

**Comment by metzm** [comment]
Explanation for my suggested changes:
Offset is only relevant if not zero. Scale is only relevant if not 1.
PR approval comes after your comments to my suggestions.

The motivation for my suggestions is to not clutter metadata with irrelevant information.

How about the need to set different offset and scale parameters for different bands in case a group is exported? I guess there is no such need, most probably you also don't see such a need.

**Comment by landam** [comment]
> Offset is only relevant if not zero. Scale is only relevant if not 1. PR approval comes after your comments to my suggestions.
> The motivation for my suggestions is to not clutter metadata with irrelevant information.

OK, make sense.

> How about the need to set different offset and scale parameters for different bands in case a group is exported? I guess there is no such need, most probably you also don't see such a need.

You are right, it's limitation of the current implementation.

**Review by metzm — APPROVED** [review]
Looks good. Let's see if there is a need for different offset/scale values per band.

**Pull request merged**